test-utils/state-viewer/src/state_dump.rs

```diff
@@ -0,0 +1,201 @@
+    use near_client::test_utils::TestEnv;
+    use near_crypto::{InMemorySigner, KeyType};
+    use near_primitives::transaction::SignedTransaction;
+    use near_primitives::types::NumBlocks;
+    use near_store::test_utils::create_test_store;
+    use near_store::Store;
+    use neard::config::GenesisExt;
+    use neard::config::TESTING_INIT_STAKE;
+    use neard::genesis_validate::validate_genesis;
+    use neard::NightshadeRuntime;
+    use std::collections::HashSet;
+    use std::iter::FromIterator;
+    use std::path::Path;
+    use std::sync::Arc;
+
+    fn setup(epoch_length: NumBlocks) -> (Arc<Store>, Genesis, TestEnv) {
+        let mut genesis = Genesis::test(vec!["test0", "test1"], 1);
+        genesis.config.num_block_producer_seats = 2;
+        genesis.config.num_block_producer_seats_per_shard = vec![2];
+        genesis.config.epoch_length = epoch_length;
+        let store = create_test_store();
+        let nightshade_runtime = NightshadeRuntime::new(
+            Path::new("."),
+            store.clone(),
+            Arc::new(genesis.clone()),
+            vec![],
+            vec![],
+        );
+        let runtimes: Vec<Arc<dyn RuntimeAdapter>> = vec![Arc::new(nightshade_runtime)];
+        let mut chain_genesis = ChainGenesis::test();
+        chain_genesis.epoch_length = epoch_length;
+        chain_genesis.gas_limit = genesis.config.gas_limit;
+        let env = TestEnv::new_with_runtime(chain_genesis, 1, 2, runtimes);
+        (store, genesis, env)
+    }
+
+    /// Test that we preserve the validators from the epoch of the state dump.
+    #[test]
+    fn test_dump_state_preserve_validators() {
+        let epoch_length = 4;
+        let (store, genesis, mut env) = setup(epoch_length);
+        let genesis_hash = env.clients[0].chain.genesis().hash();
+        let signer = InMemorySigner::from_seed("test1", KeyType::ED25519, "test1");
+        let tx = SignedTransaction::stake(
+            1,
+            "test1".to_string(),
+            &signer,
+            TESTING_INIT_STAKE,
+            signer.public_key.clone(),
+            genesis_hash,
+        );
+        env.clients[0].process_tx(tx, false);
+        for i in 1..=epoch_length * 2 + 1 {
+            env.produce_block(0, i);
+        }
+        let head = env.clients[0].chain.head().unwrap();
+        let last_block_hash = head.last_block_hash;
+        let cur_epoch_id = head.epoch_id;
+        let block_producers = env.clients[0]
+            .runtime_adapter
+            .get_epoch_block_producers_ordered(&cur_epoch_id, &last_block_hash)
+            .unwrap();
+        assert_eq!(
+            block_producers.into_iter().map(|(r, _)| r.account_id).collect::<HashSet<_>>(),
+            HashSet::from_iter(vec!["test0".to_string(), "test1".to_string()])
+        );
+        let last_block = env.clients[0].chain.get_block(&head.last_block_hash).unwrap().clone();
+        let mut state_roots = vec![];
+        for chunk in last_block.chunks.iter() {
```

Comment:
`let state_roots = last_block.chunks.iter().map(|chunk| chunk.inner.prev_state_root.clone()).collect()` ?

also given you are already passing last_block.header and we need state roots so need chunks anyway, can pass block itself and extract roots there instead of doing it on every call?

Reply:
This is in tests. In main we first computed state roots in `load_trie` so there is no reason to recompute them.